iavl/iavl_proof2.go

```diff
@@ -0,0 +1,268 @@
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/tendermint/go-wire"
+)
+
+type PathToKey struct {
+	LeafHash   []byte
+	InnerNodes []IAVLProofInnerNode
+}
+
+func (p *PathToKey) String() string {
+	str := ""
+	for i := len(p.InnerNodes) - 1; i >= 0; i-- {
+		str += p.InnerNodes[i].String() + "\n"
+	}
+	str += fmt.Sprintf("hash(leaf)=%x\n", p.LeafHash)
+	return str
+}
+
+func (p *PathToKey) Verify(leafNode IAVLProofLeafNode, root []byte) error {
+	leafHash := leafNode.Hash()
+	if !bytes.Equal(leafHash, p.LeafHash) {
+		return errors.Errorf("leaf hash does not match %x != %x", leafHash, p.LeafHash)
+	}
+	hash := leafHash
+	for _, branch := range p.InnerNodes {
+		hash = branch.Hash(hash)
+	}
+	if !bytes.Equal(root, hash) {
+		return errors.New("path does not match supplied root")
+	}
+	return nil
+}
+
+func (p *PathToKey) IsLeftmost() bool {
+	for _, node := range p.InnerNodes {
+		if node.Left != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (p *PathToKey) IsRightmost() bool {
+	for _, node := range p.InnerNodes {
+		if node.Right != nil {
+			return false
+		}
+	}
+	return true
+}
+
+type KeyExistsProof struct {
+	PathToKey
+	RootHash []byte
+}
+
+func (proof *KeyExistsProof) Verify(key []byte, value []byte, root []byte) error {
+	if !bytes.Equal(proof.RootHash, root) {
+		return errors.New("roots are not equal")
+	}
+	leafNode := IAVLProofLeafNode{KeyBytes: key, ValueBytes: value}
+	return proof.PathToKey.Verify(leafNode, root)
+}
+
+type KeyNotExistsProof struct {
+	RootHash []byte
+
+	LeftPath *PathToKey
+	LeftNode IAVLProofLeafNode
+
+	RightPath *PathToKey
+	RightNode IAVLProofLeafNode
+}
+
+func (p *KeyNotExistsProof) String() string {
+	return fmt.Sprintf("KeyNotExistsProof\nroot=%x\nleft=%s%#v\nright=%s%#v\n", p.RootHash, p.LeftPath, p.LeftNode, p.RightPath, p.RightNode)
+}
+
+func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
+	if !bytes.Equal(proof.RootHash, root) {
+		return errors.New("roots do not match")
+	}
+
+	if proof.LeftPath == nil && proof.RightPath == nil {
+		return errors.New("at least one path must exist")
+	}
+	if proof.LeftPath != nil {
+		if err := proof.LeftPath.Verify(proof.LeftNode, root); err != nil {
+			return errors.New("failed to verify left path")
+		}
+		if bytes.Compare(proof.LeftNode.KeyBytes, key) != -1 {
+			return errors.New("left node key must be lesser than supplied key")
+		}
+	}
+	if proof.RightPath != nil {
+		if err := proof.RightPath.Verify(proof.RightNode, root); err != nil {
+			return errors.New("failed to verify right path")
+		}
+		if bytes.Compare(proof.RightNode.KeyBytes, key) != 1 {
+			return errors.New("left node key must be greater than supplied key")
+		}
+	}
+
+	// Both paths exist, check that they are sequential.
+	if proof.RightPath != nil && proof.LeftPath != nil {
+		lns := proof.LeftPath.InnerNodes[:]
+		rns := proof.RightPath.InnerNodes[:]
+
+		for bytes.Compare(lns[len(lns)-1].Left, rns[len(rns)-1].Left) == 0 &&
```

Comment:
I think this could be cleaner somehow without all the `lns[len(lns)-1]`... stuff but have no concrete idea how... maybe I'll add one. I know you don't like it either.